### PR TITLE
feat: Events API

### PR DESCRIPTION
**Pull request by cuebit**
#### Type of PR:

- [ ] Bugfix
- [x] Feature
- [ ] Refactor
- [ ] Code style update
- [ ] Build-related changes
- [ ] Test
- [ ] Documentation
- [ ] Other, please describe:

#### Breaking changes:

- [x] No
- [ ] Yes

### Details

The Events API opens doors for better application hooks. This simple yet intelligent API is designed with Typescript in mind. It attempts to solve, but not limited to, the following pitfalls of a typical hook system:

- **Type-safety** – exact typings for event payloads with inference `new Events<MyEvents>`
- **Functional** – does not rely on `this` context
- **Extensive** – can be inherited by classes or used as a singleton
- **Performant** – written to execute at high operation cycles
- **Slim** – less API pollution equals more agility

### Usage

Define event interfaces using keys & argument tuples.
```ts
interface QueryEvents {
  selected: [Collection<Model>, string]
  updating: [Model]
}
```

Instantiate the Events object with the events interface.

```ts
const events = new Events<QueryEvents>()
```

Add or remove listeners.

```ts
const listener = (...args) => console.log(args)

events.on('selected', listener)
events.off('updating', listener)

// `on` returns a self-removing function.
const unregister = events.on('updating', (...args) => console.log(args))
unregister()
```

Add one-time listeners which are removed on the next broadcast event.

```ts
events.once('updating', (...args) => console.log(args))
```

Subscribe to all events.

```ts
events.subscribe(({ event, args }) => {
  if (event === 'selected') {
    const [models, entity] = args
    console.log(`${event} fired with ${models.length} for ${entity}`)
  }
})
```

Emit events.

```ts
events.emit('selected', collection, 'users')
events.emit('updating', model)
```

### Noteworthy

The Events API is designed to emit events with payloads to listeners and subscribers. If the API is offered to authors within the ecosystem and general user-land, performance and optimisation plays a key role as the use-cases can be unprecedented. To cater for these needs, the following applies:

- event listener maps are initialized with `Object.create(null)` to increase operation cycles.
- event keys are removed to optimize memory when listeners for an event are unregistered and no more listeners exist.
- listeners are not required to return which results in less operating overhead.
- `removeAllListeners(event: string)` is a protected method to prevent side-effects from user-land intervention.
- recursion and advanced usage is demonstrated within tests.

### CODE REVIEW

File: src/events/Events.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+   * Creates an Events instance.
+   */
+  constructor() {
+    this.listeners = Object.create(null)
+    this.subscribers = []
+  }
+
+  /**
+   * Register a listener for a given event.
+   * @returns a function that, when called, will unregister the handler.
```
[the remark this answers is not in the document]
Comment (cuebit) [reply src/events/Events.ts]: Done 👍

File: src/events/Events.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+   */
+  on<K extends keyof T>(event: K, callback: EventListener<T, K>): () => void {
+    if (!event || !isFunction(callback)) {
+      return () => {} // Non-blocking noop
```
[the remark this answers is not in the document]
Comment (cuebit) [reply src/events/Events.ts]: Done 👍

**Comment by kiaking** [comment]
One another thing I want to discuss is how we get the results from the `emit` method. As for event API, it doesn't make much sense for `emit` to return a value since event can call event within event and such.

However, in Vuex ORM hooks, we need to know what user have returned. For example, if users return `false` from the `beforeCreate` hook, it should stop persisting it.

This discussion could be in different issue since this event api is pretty solid, though do you have any idea on this topic?

File: src/events/Events.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+export type EventSubscriberArgs<T> = {
+  [K in keyof T]: { event: K; args: EventArgs<T[K]> }
+}[keyof T]
+
+export type EventSubscriber<T> = (arg: EventSubscriberArgs<T>) => void
+
+/**
+ * Events class for listening to and emitting of events.
+ * @public
```
Comment (kiaking) [inline src/events/Events.ts:17]: Oops sorry, blank line here too 🙏

**Comment by cuebit** [comment]
I’ve opened this up for debate here #12.

**Comment by kiaking** [comment]
Great. Let's follow up the discussion over there. I'm gonna merge this since this looks pretty tasty 🥰

**Pull request merged**